File: Solutions/Array-String/27-Remove-Element/csharp/Solution_v1.cs
/*
Problem: 27. Remove Element
Difficulty: Easy
Link: https://leetcode.com/problems/remove-element/
Approach: Naive - Two Pointers
Time Complexity: O(n)
Space Complexity: O(1)
*/

public class Solution {
    public int RemoveElement(int[] nums, int val) {
        int k = nums.Length - 1;
        int i = 0;

        while (i <= k)
        {
            if (nums[i] == val)
            {
                nums[i] = nums[k--];
            }
            else
            {
                i++;
            }
        }

        return k + 1;
    }
}

File: Solutions/Array-String/27-Remove-Element/go/Solution_v1.go
// Problem: 27. Remove Element
// Difficulty: Easy
// Link: https://leetcode.com/problems/remove-element/
// Approach: Two Pointers
// Time Complexity: O(n)
// Space Complexity: O(1)

func removeElement(nums []int, val int) int {
    k, p := len(nums) - 1, 0

    for p <= k {
        if nums[p] == val {
            nums[p] = nums[k]
            k--
        } else {
            p++
        }
    }

    return k + 1
}
